src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -399,12 +399,17 @@ private BoundPattern BindDiscardPattern(DiscardPatternSyntax node, TypeSymbol in
                 diagnostics.Add(ErrorCode.ERR_PointerTypeInPatternMatching, typeSyntax.Location);
                 return true;
             }
-            else if (patternType.IsNullableType() || typeSyntax is NullableTypeSyntax)
+            else if (patternType.IsNullableType())
             {
                 // It is an error to use pattern-matching with a nullable type, because you'll never get null. Use the underlying type.
                 Error(diagnostics, ErrorCode.ERR_PatternNullableType, typeSyntax, patternType, patternType.GetNullableUnderlyingType());
                 return true;
             }
+            else if (typeSyntax is NullableTypeSyntax)
+            {
+                Error(diagnostics, ErrorCode.ERR_PatternNullableType, typeSyntax, patternType.ToDisplayString(CodeAnalysis.NullableFlowState.MaybeNull), patternType);
```

Comment:
Ideally we should delay converting the `patternType` argument to a string until the diagnostic message is needed. Could we use `typeSyntax` as the diagnostic argument instead of `patternType.ToDisplayString(...)`?

Reply:
No, because we are using `ToDisplayString` to tack on a `?`. I suppose I could put the `?` into the diagnostic template?

---
In reply to: [440484936](https://github.com/dotnet/roslyn/pull/45206#discussion_r440484936) [](ancestors = 440484936)

Reply:
`typeSyntax` is a `NullableTypeSyntax`. Doesn't it include `?`?

Reply:
`typeSyntax` is for the squiggles.

---
In reply to: [440552553](https://github.com/dotnet/roslyn/pull/45206#discussion_r440552553) [](ancestors = 440552553)